src/Syntax/Types.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE MultiParamTypeClasses, FunctionalDependencies #-}
+
+-- These extensions should be removed once we remove the defs at the end of this file.
+{-# LANGUAGE RankNTypes, TypeApplications, StandaloneKindSignatures, PolyKinds #-}
+
+{-|
+Module      : Syntax.Types
+Description : Types that represent Act types, and functions and patterns to go between them and Haskell's own types.
+-}
+
+module Syntax.Types (module Syntax.Types) where
+
+import Data.Singletons
+import Data.Type.Equality (TestEquality(..))
+import Data.Typeable hiding (TypeRep,typeRep)
+import Type.Reflection
+
+import Data.ByteString    as Syntax.Types (ByteString)
+import EVM.ABI            as Syntax.Types (AbiType(..))
+
+-- | Types understood by proving tools.
+data MType
```

Comment:
Good question actually! I wondered the same thing myself for quite a while. In the end I concluded that it needs to be there but now that you forced me to think about it again, I realized that it's unnecessary.

It *is* necessary if we want to define an instance of the `SingKind` class, but we actually don't need to do that. We only go from `AbiType` to `SType` (sometimes via `MType`), but never the other way around. So we don't need a class that lets us go back from `SType`.

I'll push a change tonight where `MType` is still left, but has been turned into a type synonym instead of remaining as a proper type of its own. This lets us leave the rest of the code base completely unaffected, while essentially removing it. We can remove it completely later, if we want.

Reply:
> We can remove it completely later, if we want.

Nice, this should be a fairly mechanical case of find / replace iiuc?

Reply:
Yes, just `s/MType/SomeSing *`. But honestly I find that syntax less informative, so I'd vote against. Maybe `s/MType/ActType` or some such?